pintc/src/predicate/optimize/const_folding.rs

```diff
@@ -30,19 +66,203 @@ pub(crate) fn fold_consts(contract: &mut Contract) {
                     span: expr_key.get(contract).span().clone(),
                 };
 
-                replace_map.insert(
+                expr_keys_to_replace.insert(
                     expr_key,
                     (simplified_expr, expr_key.get_ty(contract).clone()),
                 );
             }
         }
 
-        for (old_expr_key, (simplified_expr, simplified_type)) in &replace_map {
+        for (old_expr_key, (simplified_expr, simplified_type)) in &expr_keys_to_replace {
             let simplified_expr_key = contract
                 .exprs
                 .insert(simplified_expr.clone(), simplified_type.clone());
 
             contract.replace_exprs(Some(pred_key), *old_expr_key, simplified_expr_key);
         }
+
+        if !expr_keys_to_replace.is_empty() {
+            did_fold_const = true;
+        }
+        expr_keys_to_replace = FxHashMap::default();
     }
+
+    did_fold_const
+}
+
+/// In a given contract, simplify any binary ops that consist of boolean or arithmetic (real and int) identities.
+/// Returns whether or not any identities were folded.
+///
+/// Ex:
+/// true || <expr> is true
+/// true && <expr> is <expr>
+/// false || <expr> is <expr>
+/// false && <expr> is <false>
+/// x + 0 is x
+/// x - 0 is x
+/// x * 0 is 0
+/// x * 1 is x
+/// 0 / x is 0
+/// x / 1 is x
+/// x % 1 is x
+pub(crate) fn fold_identities(contract: &mut Contract) -> bool {
+    let mut expr_keys_to_replace: Vec<(ExprKey, ExprKey)> = vec![];
+    let mut did_fold_const: bool = false;
+
+    for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        for expr_key in contract.exprs(pred_key) {
+            if let Expr::BinaryOp { op, lhs, rhs, .. } = expr_key.get(contract) {
+                let lhs_imm = if let Expr::Immediate { value, .. } = lhs.get(contract) {
+                    Some(value)
+                } else {
+                    None
+                };
+
+                let rhs_imm = if let Expr::Immediate { value, .. } = rhs.get(contract) {
+                    Some(value)
+                } else {
+                    None
+                };
+
+                match op {
+                    BinaryOp::LogicalAnd => match (lhs_imm, rhs_imm) {
+                        (Some(Immediate::Bool(true)), _) => {
+                            expr_keys_to_replace.push((expr_key, *rhs));
+                        }
```

Comment:
Might be easier to consolidate the `match` statements?
```
                match (op, lhs_imm, rhs_imm) {
                    // Boolean identities
                    (BinaryOp::LogicalAnd, Some(Immediate::Bool(true)), _) => Some(*rhs),
                    (BinaryOp::LogicalAnd, _, Some(Immediate::Bool(true))) => Some(*lhs),
                    ...
                    // Arithmetic identities
                    (BinaryOp::Add, Some(Immediate::Int(0)), _) | (BinaryOp::Add, _, Some(Immediate::Int(0))) => Some(*rhs),
                    (BinaryOp::Add, Some(Immediate::Real(0.0)), _) | (BinaryOp::Add, _, Some(Immediate::Real(0.0))) => Some(*rhs),
```

Reply:
Something like this..

Reply:
Updated